### PR TITLE
feat: error if output fails validation according to global setting

### PR DESCRIPTION
**Pull request by beckermr**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by beckermr** [comment]
@isuruf @conda-forge/core this is ready for review!

File: recipe/conda_forge_ci_setup/feedstock_outputs.py [hunk 1]
```diff
@@ -10,10 +10,28 @@
 from .utils import built_distributions, compute_sha256sum, split_pkg
 
 
-VALIDATION_ENDPOINT = "https://conda-forge.herokuapp.com"
-STAGING = "cf-staging"
+@lru_cache(maxsize=1)
```
Comment (beckermr) [inline recipe/conda_forge_ci_setup/feedstock_outputs.py:13]:
```suggestion
VALIDATION_ENDPOINT = "https://conda-forge.herokuapp.com"
STAGING = "cf-staging"


@lru_cache(maxsize=1)
```

File: recipe/conda_forge_ci_setup/feedstock_outputs.py [hunk 2]
```diff
@@ -14,6 +15,28 @@
 STAGING = "cf-staging"
 
 
+@lru_cache(maxsize=1)
+def _load_allowed_autoreg_feedstock_globs(time_int):
+    r = requests.get(
+        "https://raw.githubusercontent.com/conda-forge/admin-requests/"
+        "main/.feedstock_outputs_autoreg_allowlist.yml"
```
Comment (beckermr) [inline recipe/conda_forge_ci_setup/feedstock_outputs.py:22]:
```suggestion
        "https://raw.githubusercontent.com/conda-forge/feedstock-outputs/"
        "main/feedstock_outputs_autoreg_allowlist.yml"
```

File: recipe/conda_forge_ci_setup/feedstock_outputs.py [hunk 3]
```diff
@@ -104,7 +127,7 @@ def is_valid_feedstock_output(project, outputs):
             except requests.HTTPError as exc:
                 if exc.response.status_code == 404:
                     # no output exists and we can add it
-                    valid[dist] = True
+                    valid[dist] = True or check_allowed_autoreg_feedstock_globs(feedstock, o)
```
Comment (beckermr) [inline recipe/conda_forge_ci_setup/feedstock_outputs.py:130]:
```suggestion
                    # if we turn off autoregistration, then the True here should be False
                    valid[dist] = True or check_allowed_autoreg_feedstock_globs(feedstock, o)
```

File: recipe/conda_forge_ci_setup/feedstock_outputs.py [hunk 4]
```diff
@@ -14,6 +17,28 @@
...
+    return _load_allowed_autoreg_feedstock_globs(time.monotonic() // 120)
+
+
+def check_allowed_autoreg_feedstock_globs(feedstock, output):
+    fs_pats = load_allowed_autoreg_feedstock_globs()
+    for pat in fs_pats.get(feedstock, []):
+        if fnmatch(output, pat):
+            return True
+    return False
+
```
Comment (beckermr) [inline recipe/conda_forge_ci_setup/feedstock_outputs.py:41]:
```suggestion


```

File: recipe/conda_forge_ci_setup/feedstock_outputs.py [hunk 5]
```diff
@@ -14,6 +17,29 @@
...
+def load_allowed_autoreg_feedstock_globs():
+    return _load_allowed_autoreg_feedstock_globs(time.monotonic() // 120)
+
+
+def check_allowed_autoreg_feedstock_globs(feedstock, output):
+    fs_pats = load_allowed_autoreg_feedstock_globs()
+    for pat in fs_pats.get(feedstock, []):
+        if fnmatch(output, pat):
+            return True
+    return False
```
Comment (isuruf) [inline recipe/conda_forge_ci_setup/feedstock_outputs.py:40]: Shouldn't these go in https://github.com/conda-forge/conda-forge-metadata/blob/main/conda_forge_metadata/feedstock_outputs.py ?

Comment (beckermr) [reply recipe/conda_forge_ci_setup/feedstock_outputs.py:40]: Sure, I can move them there.

**Comment by beckermr** [comment]
@conda-forge-admin restart ci

File: recipe/meta.yaml [hunk 6]
```diff
@@ -80,7 +81,7 @@ requirements:
     - m2-git  # [win]
     - git     # [unix]
     - libarchive
-    - conda-forge-metadata >=0.4.1
+    - conda-forge-metadata >=0.9.0
```
Comment (beckermr) [inline recipe/meta.yaml:84]:
```suggestion
    - conda-forge-metadata >=0.9.1
```

**Comment by beckermr** [comment]
@conda-forge-admin restart ci

**Comment by beckermr** [comment]
ok finally got all the bugs - ready for another look @isuruf @conda-forge/core!

**Comment by beckermr** [comment]
ping here again @isuruf

File: recipe/conda_forge_ci_setup/feedstock_outputs.py [hunk 7]
```diff
@@ -97,19 +98,20 @@ def is_valid_feedstock_output(project, outputs):
...
-        
+
         for i in range(3):  # three attempts
             try:
                 registered_feedstocks = package_to_feedstock(o)
             except requests.HTTPError as exc:
                 if exc.response.status_code == 404:
                     # no output exists and we can add it
+                    # if we turn off autoregistration, then the True here should be False
                     valid[dist] = True
```
Comment (beckermr) [inline recipe/conda_forge_ci_setup/feedstock_outputs.py]: @isuruf This PR does not switch the default yet. If you'd rather have it do that and wait, that is fine.

Comment (beckermr) [inline recipe/conda_forge_ci_setup/feedstock_outputs.py]:
```suggestion
                    valid[dist] = False  # do not change to true!
```

**Comment by isuruf** [comment]
I meant, since we are still allowing auto-registration, we shouldn't error out for now.

**Comment by beckermr** [comment]
Agreed! I change this pr to do the full thing and will not merge yet.

**Comment by isuruf** [comment]
We can also add a `ALLOW_AUTOREG = True` constant at the top and error out if it is False. So it is just a toggling that's needed.

**Comment by beckermr** [comment]
It'd be nice to have a global setting somewhere to be honest. Maybe we add it to feedstock outputs and then have both the webservices and this package pull the setting on the fly.

File: recipe/conda_forge_ci_setup/feedstock_outputs.py [hunk 8]
```diff
@@ -97,19 +101,19 @@ def is_valid_feedstock_output(project, outputs):
...
+
         for i in range(3):  # three attempts
             try:
                 registered_feedstocks = package_to_feedstock(o)
-            except requests.HTTPError as exc:
+            except requests.exceptions.HTTPError as exc:
                 if exc.response.status_code == 404:
-                    # no output exists and we can add it
-                    valid[dist] = True
+                    # no output exists so see if we can add itand we can add it
```
Comment (beckermr) [inline recipe/conda_forge_ci_setup/feedstock_outputs.py:110]:
```suggestion
                    # no output exists so see if we can add it
```

**Comment by beckermr** [comment]
woops I hit the wrong button!

**Comment by beckermr** [comment]
This was ready to merge and the tests were passing, so if it works, i'll leave it. Otherwise, I'll clean up any buggy outputs.

**Comment by JohanMabille** [comment]
It looks like this makes the micromamba feedstock fail on osx-arm64 (at least when built locally) with the following error:
```
Validating outputs
+ validate_recipe_outputs ''
validation results:
{
  "osx-arm64/micromamba-2.0.0rc4-1.tar.bz2": false
}
NOTE: Any outputs marked as False are not allowed for this feedstock. See https://conda-forge.org/docs/maintainer/infrastructure/#output-validation-and-feedstock-tokens for information on how to address this error.
valid configs are {'osx_arm64_', 'linux_aarch64_', 'win_64_', 'linux_ppc64le_', 'linux_64_', 'osx_64_'}
Traceback (most recent call last):
  File "/Users/runner/work/mamba/mamba/micromamba-feedstock/build-locally.py", line 109, in <module>
Using osx_arm64_ configuration
    main()
  File "/Users/runner/work/mamba/mamba/micromamba-feedstock/build-locally.py", line 99, in main
    run_osx_build(ns)
  File "/Users/runner/work/mamba/mamba/micromamba-feedstock/build-locally.py", line 37, in run_osx_build
    subprocess.check_call([script])
  File "/Users/runner/micromamba/envs/mambabuild/lib/python3.12/subprocess.py", line 413, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['.scripts/run_osx_build.sh']' returned non-zero exit status 1.
```

I don't know how to fix the feedstock.

**Comment by beckermr** [comment]
Can you retry the build?

**Comment by beckermr** [comment]
Also a link to the ci job would be helpful.

**Comment by JohanMabille** [comment]
Ah sorry for that, here is the [ci job](https://github.com/mamba-org/mamba/actions/runs/10808207917/job/29989551546?pr=3434), I've just restarted it.

**Comment by beckermr** [comment]
Ahhh this is outside of conda-forge. We can turn this off there.

**Pull request merged**